sdk/core/azure-core/azure/core/polling/base_polling.py

```diff
@@ -0,0 +1,552 @@
+    """Check response status code is valid.
+
+    Must be 200, 201, 202, or 204.
+
+    :raises: BadStatus if invalid status.
+    """
+    code = response.status_code
+    if code in {200, 201, 202, 204}:
+        return
+    raise BadStatus(
+        "Invalid return status {!r} for {!r} operation".format(
+            code, response.request.method
+        )
+    )
+
+
+def _is_empty(response):
+    # type: (ResponseType) -> bool
+    """Check if response body contains meaningful content.
+
+    :rtype: bool
+    :raises: DecodeError if response body contains invalid json data.
+    """
+    content = response.text()
+    if not content:
+        return True
+    try:
+        return not json.loads(content)
+    except ValueError:
+        raise DecodeError("Error occurred in deserializing the response body.")
+
+
+class LongRunningOperation(ABC):
+    """LongRunningOperation
+    Provides default logic for interpreting operation responses
+    and status updates.
+
+    :param azure.core.pipeline.PipelineResponse response: The initial pipeline response.
+    :param callable deserialization_callback: The deserialization callaback.
+    :param dict lro_options: LRO options.
+    :param kwargs: Unused for now
+    """
+
+    @abc.abstractmethod
+    def can_poll(self, pipeline_response):
+        # type: (PipelineResponseType) -> bool
+        """Answer if this polling method could be used.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_polling_url(self):
+        """Return the polling URL.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def set_initial_status(self, pipeline_response):
+        # type: (PipelineResponseType) -> str
+        """Process first response after initiating long running operation.
+
+        :param azure.core.pipeline.PipelineResponse response: initial REST call response.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_status(self, pipeline_response):
+        # type: (PipelineResponseType) -> str
+        """Return the status string extracted from this response."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_final_get_url(self, pipeline_response):
+        # type: (PipelineResponseType) -> Optional[str]
+        """If a final GET is needed, returns the URL.
+
+        :rtype: str
+        """
+        raise NotImplementedError()
+
+
+class OperationResourcePolling(LongRunningOperation):
+    """Implements a operation resource polling, typically from Operation-Location.
+
+    :param str operation_location_header: Name of the header to return operation format (default 'operation-location')
+    """
+
+    def __init__(self, operation_location_header="operation-location"):
+        self._operation_location_header = operation_location_header
+
+        # Store the initial URLs
+        self.async_url = None
+        self.location_url = None
+        self.request = None
```

Comment:
Do we want to expose these members?

It looks to me we prefer to let users to access the members via get_polling_url & get_final_get_url?

Reply:
fixed